activerecord/lib/active_record/associations/has_one_association.rb

```diff
@@ -56,12 +56,18 @@ def replace(record, save = true)
 
           return target unless load_target || record
 
-          assigning_another_record = target != record
+          record_foreign_key_changed = record && record.changes_to_save.include?(reflection.foreign_key)
```

Comment:
There is a special method to check if the attribute will change on save:
```suggestion          
          record_foreign_key_changed = record && record.will_save_change_to_attribute?(reflection.foreign_key)
```

Reply:
Thanks @p8 I've actioned this!